File: domain/src/test/java/com/github/anrimian/musicplayer/domain/interactors/player/LibraryPlayerInteractorTest.java
package com.github.anrimian.musicplayer.domain.interactors.player;

import static com.github.anrimian.musicplayer.domain.interactors.TestBusinessDataProvider.currentItem;
import static com.github.anrimian.musicplayer.domain.interactors.TestBusinessDataProvider.fakeCompositionSource;
import static com.github.anrimian.musicplayer.domain.interactors.TestBusinessDataProvider.getFakeCompositions;
import static com.github.anrimian.musicplayer.domain.models.player.error.ErrorType.NOT_FOUND;
import static com.github.anrimian.musicplayer.domain.models.player.error.ErrorType.UNKNOWN;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics;
import com.github.anrimian.musicplayer.domain.models.composition.CorruptionType;
import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.ErrorEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.FinishedEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PreparedEvent;
import com.github.anrimian.musicplayer.domain.models.player.modes.RepeatMode;
import com.github.anrimian.musicplayer.domain.repositories.LibraryRepository;
import com.github.anrimian.musicplayer.domain.repositories.PlayQueueRepository;
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository;
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.Mockito;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class LibraryPlayerInteractorTest {

    private final PlayerCoordinatorInteractor playerCoordinatorInteractor = mock(PlayerCoordinatorInteractor.class);
    private final SettingsRepository settingsRepository = mock(SettingsRepository.class);
    private final PlayQueueRepository playQueueRepository = mock(PlayQueueRepository.class);
    private final LibraryRepository musicProviderRepository = mock(LibraryRepository.class);
    private final UiStateRepository uiStateRepository = mock(UiStateRepository.class);
    private final Analytics analytics = mock(Analytics.class);

    private LibraryPlayerInteractor libraryPlayerInteractor;

    private final PublishSubject<PlayerEvent> playerEventSubject = PublishSubject.create();
    private final BehaviorSubject<PlayQueueEvent> currentCompositionSubject = BehaviorSubject.createDefault(currentItem(0));

    private final InOrder inOrder = Mockito.inOrder(playQueueRepository,
            playerCoordinatorInteractor,
            musicProviderRepository);

    @Before
    public void setUp() {
        when(playQueueRepository.setPlayQueue(any())).thenReturn(Completable.complete());
        when(playQueueRepository.setPlayQueue(any(), anyInt())).thenReturn(Completable.complete());
        when(playQueueRepository.getCurrentQueueItemObservable())
                .thenReturn(currentCompositionSubject);
        when(playQueueRepository.skipToNext()).thenReturn(Single.just(1));

        when(playerCoordinatorInteractor.getPlayerEventsObservable(any())).thenReturn(playerEventSubject);
        when(playerCoordinatorInteractor.getActualTrackPosition(any())).thenReturn(Single.just(0L));

        when(musicProviderRepository.writeErrorAboutComposition(any(), any()))
                .thenReturn(Completable.complete());
        when(musicProviderRepository.deleteComposition(any())).thenReturn(Completable.complete());

        when(settingsRepository.isDecreaseVolumeOnAudioFocusLossEnabled()).thenReturn(true);
        when(settingsRepository.isPauseOnAudioFocusLossEnabled()).thenReturn(true);

        libraryPlayerInteractor = new LibraryPlayerInteractor(
                playerCoordinatorInteractor,
                settingsRepository,
                playQueueRepository,
                musicProviderRepository,
                uiStateRepository,
                analytics
        );
    }

    @Test
    public void startPlayingTest() {
        libraryPlayerInteractor.startPlaying(getFakeCompositions(), 0);

        verify(playQueueRepository).setPlayQueue(getFakeCompositions(), 0);
        verify(playerCoordinatorInteractor).prepareToPlay(eq(fakeCompositionSource(0)), any());
    }

    @Test
    public void playWithoutPreparingTest() {
        libraryPlayerInteractor.play();

        verify(playerCoordinatorInteractor).prepareToPlay(any(), any());
    }

    @Test
    public void onPlayFinishedTest() {
        libraryPlayerInteractor.play();

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new FinishedEvent(composition));
        inOrder.verify(playQueueRepository).skipToNext();

        currentCompositionSubject.onNext(currentItem(1));
        CompositionSource nextComposition = fakeCompositionSource(1);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(nextComposition), any());

        playerEventSubject.onNext(new PreparedEvent(nextComposition));
    }

    @Test
    public void onPlayToEndTest() {
        when(playQueueRepository.skipToNext()).thenReturn(Single.just(0));

        libraryPlayerInteractor.play();

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new FinishedEvent(composition));
        inOrder.verify(playQueueRepository).skipToNext();

        currentCompositionSubject.onNext(currentItem(1));
        inOrder.verify(playerCoordinatorInteractor).stop(any());
    }

    @Test
    public void onPlayToEndWithRepeatPlayListModeTest() {
        when(playQueueRepository.skipToNext()).thenReturn(Single.just(0));
        when(settingsRepository.getRepeatMode()).thenReturn(RepeatMode.REPEAT_PLAY_LIST);

        libraryPlayerInteractor.play();

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new FinishedEvent(composition));
        inOrder.verify(playQueueRepository).skipToNext();

        currentCompositionSubject.onNext(currentItem(1));
        CompositionSource nextComposition = fakeCompositionSource(1);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(nextComposition), any());
    }

    @Test
    public void onCompositionPlayFinishedInStopState() {
        libraryPlayerInteractor.play();
        libraryPlayerInteractor.stop();

        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(any(), any());
        inOrder.verify(playerCoordinatorInteractor).stop(any());

        CompositionSource composition = fakeCompositionSource(0);
        playerEventSubject.onNext(new FinishedEvent(composition));

        inOrder.verify(playQueueRepository).skipToNext();

        currentCompositionSubject.onNext(currentItem(1));

        CompositionSource nextComposition = fakeCompositionSource(1);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(nextComposition), any());
    }

    @Test
    public void onCompositionUnknownErrorTest() {
        libraryPlayerInteractor.play();

        when(playQueueRepository.isCurrentCompositionAtEndOfQueue()).thenReturn(Single.just(false));

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new ErrorEvent(UNKNOWN, composition));

        inOrder.verify(musicProviderRepository)
                .writeErrorAboutComposition(CorruptionType.UNKNOWN, getFakeCompositions().get(0));
        inOrder.verify(playQueueRepository).skipToNext();

        currentCompositionSubject.onNext(currentItem(1));

        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(fakeCompositionSource(1)), any());
    }

    @Test
    public void onCompositionUnknownErrorInEndTest() {
        when(playQueueRepository.isCurrentCompositionAtEndOfQueue()).thenReturn(Single.just(false));

        libraryPlayerInteractor.play();

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new ErrorEvent(UNKNOWN, composition));

        inOrder.verify(musicProviderRepository)
                .writeErrorAboutComposition(CorruptionType.UNKNOWN, getFakeCompositions().get(0));
        inOrder.verify(playQueueRepository).skipToNext();

        when(playQueueRepository.isCurrentCompositionAtEndOfQueue()).thenReturn(Single.just(true));

        currentCompositionSubject.onNext(currentItem(1));
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(fakeCompositionSource(1)), any());

        playerEventSubject.onNext(new ErrorEvent(UNKNOWN, composition));
        inOrder.verify(playerCoordinatorInteractor, never()).play(any(), anyInt());
    }

    @Test
    public void onCompositionDeletedErrorTest() {
        when(playQueueRepository.isCurrentCompositionAtEndOfQueue()).thenReturn(Single.just(false));

        libraryPlayerInteractor.play();

        CompositionSource composition = fakeCompositionSource(0);
        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(composition), any());

        playerEventSubject.onNext(new ErrorEvent(NOT_FOUND, composition));

        inOrder.verify(musicProviderRepository)
                .writeErrorAboutComposition(CorruptionType.NOT_FOUND, getFakeCompositions().get(0));
        inOrder.verify(playQueueRepository).skipToNext();
    }

    @Test
    public void onCurrentCompositionDeletedChangeTest() {
        libraryPlayerInteractor.play();

        currentCompositionSubject.onNext(new PlayQueueEvent(null));

        verify(playerCoordinatorInteractor).reset(eq(PlayerType.LIBRARY));
    }

    @Test
    public void skipToPreviousTest() {
        libraryPlayerInteractor.play();

        inOrder.verify(playerCoordinatorInteractor).prepareToPlay(eq(fakeCompositionSource(0)), any());

        when(settingsRepository.getSkipConstraintMillis()).thenReturn(15);
        when(playerCoordinatorInteractor.getActualTrackPosition(any())).thenReturn(Single.just(10L));
        libraryPlayerInteractor.skipToPrevious();
        inOrder.verify(playQueueRepository).skipToPrevious();

        when(settingsRepository.getSkipConstraintMillis()).thenReturn(15);
        when(playerCoordinatorInteractor.getActualTrackPosition(any())).thenReturn(Single.just(30L));
        libraryPlayerInteractor.skipToPrevious();
        inOrder.verify(playerCoordinatorInteractor).onSeekFinished(eq(0L), any());
    }

}

File: domain/src/test/java/com/github/anrimian/musicplayer/domain/interactors/player/LibraryPlayerInteractorTest.kt
package com.github.anrimian.musicplayer.domain.interactors.player

import com.github.anrimian.filesync.SyncInteractor
import com.github.anrimian.musicplayer.domain.controllers.SystemServiceController
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics
import com.github.anrimian.musicplayer.domain.models.composition.CorruptionType
import com.github.anrimian.musicplayer.domain.models.composition.content.AcceptablePlayerException
import com.github.anrimian.musicplayer.domain.models.composition.source.LibraryCompositionSource
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem
import com.github.anrimian.musicplayer.domain.models.player.PlayerState
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent
import com.github.anrimian.musicplayer.domain.models.player.modes.RepeatMode
import com.github.anrimian.musicplayer.domain.models.sync.FileKey
import com.github.anrimian.musicplayer.domain.repositories.LibraryRepository
import com.github.anrimian.musicplayer.domain.repositories.PlayQueueRepository
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject
import io.reactivex.rxjava3.subjects.PublishSubject
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class LibraryPlayerInteractorTest {

    private val playerCoordinatorInteractor: PlayerCoordinatorInteractor = mock()
    private val syncInteractor: SyncInteractor<FileKey, *, Long> = mock()
    private val settingsRepository: SettingsRepository = mock()
    private val playQueueRepository: PlayQueueRepository = mock()
    private val musicProviderRepository: LibraryRepository = mock()
    private val uiStateRepository: UiStateRepository = mock()
    private val systemServiceController: SystemServiceController = mock()
    private val analytics: Analytics = mock()
    
    private lateinit var libraryPlayerInteractor: LibraryPlayerInteractor

    private val playQueueItem: PlayQueueItem = mock {
        on { duration } doReturn 60_000L
    }
    private val testLibrarySource: LibraryCompositionSource = mock {
        on { composition } doReturn playQueueItem
    }
    private val testPlayQueueEvent: PlayQueueEvent = mock {
        on { playQueueItem } doReturn playQueueItem
    }
    private var isCurrentCompositionErrorEnabled = false

    private val playerEventSubject = PublishSubject.create<PlayerEvent>()
    private val playerStateSubject = PublishSubject.create<PlayerState>()
    private val currentCompositionSubject = BehaviorSubject.createDefault(testPlayQueueEvent)
    private val trackPositionSubject = BehaviorSubject.createDefault(0L)

    private val inOrder = Mockito.inOrder(
        playerCoordinatorInteractor,
        settingsRepository,
        playQueueRepository,
        musicProviderRepository,
        uiStateRepository,
        analytics
    )

    @BeforeEach
    fun setUp() {
        whenever(playQueueRepository.setPlayQueue(any(), anyInt())).thenReturn(Completable.complete())
        whenever(playQueueRepository.getCurrentQueueItemObservable())
            .thenReturn(currentCompositionSubject.map {
                if (isCurrentCompositionErrorEnabled) {
                    throw Exception()
                }
                return@map it
            })
        whenever(playQueueRepository.getNextQueueItemId()).thenReturn(Single.just(1L))
        whenever(playQueueRepository.skipToNext()).thenReturn(Single.just(1))
        whenever(playQueueRepository.getItemTrackPosition(any())).thenReturn(Single.just(0L))
        whenever(playQueueRepository.setItemTrackPosition(any(), any())).thenReturn(Completable.complete())
        whenever(playQueueRepository.setCurrentItemTrackPosition(any())).thenReturn(Completable.complete())

        whenever(playerCoordinatorInteractor.getPlayerEventsObservable(any())).thenReturn(playerEventSubject)
        whenever(playerCoordinatorInteractor.getPlayerStateObservable(any())).thenReturn(playerStateSubject)
        whenever(playerCoordinatorInteractor.getIsPlayingStateObservable(any()))
            .thenReturn(Observable.just(false))
        whenever(playerCoordinatorInteractor.getTrackPositionObservable(any()))
            .thenReturn(trackPositionSubject)

        whenever(musicProviderRepository.writeErrorAboutComposition(any(), any()))
            .thenReturn(Completable.complete())
        whenever(musicProviderRepository.deleteComposition(any()))
            .thenReturn(Single.just(mock()))

        whenever(settingsRepository.isDecreaseVolumeOnAudioFocusLossEnabled).thenReturn(true)
        whenever(settingsRepository.isPauseOnAudioFocusLossEnabled).thenReturn(true)

        libraryPlayerInteractor = LibraryPlayerInteractor(
            playerCoordinatorInteractor,
            syncInteractor,
            settingsRepository,
            playQueueRepository,
            musicProviderRepository,
            uiStateRepository,
            systemServiceController,
            analytics
        )
    }

    @Test
    fun `start playing test`() {
        val queue: List<Long> = mock()
        libraryPlayerInteractor.setQueueAndPlay(queue, 0).subscribe()

        playQueueRepository.setPlayQueue(eq(queue), eq(0))

        verify(playerCoordinatorInteractor).playAfterPrepare(eq(PlayerType.LIBRARY))
    }

    @Test
    fun `start playing with source prepare error`() {
        val ex = Exception()
        currentCompositionSubject.onError(ex)

        val queue: List<Long> = mock()
        libraryPlayerInteractor.setQueueAndPlay(queue, 0).subscribe()

        verify(playerCoordinatorInteractor, never()).playAfterPrepare(eq(PlayerType.LIBRARY))
    }

    @Test
    fun `play, error, play again`() {
        isCurrentCompositionErrorEnabled = true
        currentCompositionSubject.onNext(testPlayQueueEvent)

        libraryPlayerInteractor.play()
        verify(playerCoordinatorInteractor, never()).play(eq(PlayerType.LIBRARY), any())
        verify(systemServiceController).stopMusicService(eq(true), eq(true))

        isCurrentCompositionErrorEnabled = false
        currentCompositionSubject.onNext(testPlayQueueEvent)
        libraryPlayerInteractor.play()
        verify(playerCoordinatorInteractor).play(eq(PlayerType.LIBRARY), any())
    }

    @Nested
    @DisplayName("without prepare")
    inner class WithoutPrepare {

        @Test
        fun `just play`() {
            libraryPlayerInteractor.play()

            verify(playerCoordinatorInteractor).play(eq(PlayerType.LIBRARY), any())
        }

        @Test
        fun `seek to`() {
            val pos = 10_000L
            libraryPlayerInteractor.onSeekFinished(pos)

            verify(playQueueRepository).setCurrentItemTrackPosition(eq(pos))
            verify(playerCoordinatorInteractor).onSeekFinished(eq(pos), eq(PlayerType.LIBRARY))
        }

        @Test
        fun `seek by`() {
            val pos = 10_000L
            whenever(settingsRepository.rewindValueMillis).thenReturn(pos)

            libraryPlayerInteractor.fastSeekForward()

            verify(playQueueRepository).setCurrentItemTrackPosition(eq(pos))
            verify(playerCoordinatorInteractor).onSeekFinished(eq(pos), eq(PlayerType.LIBRARY))
        }

    }

    @Nested
    @DisplayName("after prepare")
    inner class AfterPrepare {

        @BeforeEach
        fun setUp() {
            libraryPlayerInteractor.prepare()
        }

        @Test
        fun `on finished event received`() {
            playerEventSubject.onNext(PlayerEvent.FinishedEvent(testLibrarySource))

            verify(playQueueRepository).skipToNext()
        }

        @Test
        fun `on finished event received with last item in queue`() {
            whenever(playQueueRepository.skipToNext()).thenReturn(Single.just(0))

            playerEventSubject.onNext(PlayerEvent.FinishedEvent(testLibrarySource))

            verify(playQueueRepository).skipToNext()
            verify(playerCoordinatorInteractor).pause(eq(PlayerType.LIBRARY))
        }

        @Test
        fun `on finished event received with last item in queue with repeat mode`() {
            whenever(playQueueRepository.skipToNext()).thenReturn(Single.just(0))
            whenever(settingsRepository.repeatMode).thenReturn(RepeatMode.REPEAT_PLAY_QUEUE)

            playerEventSubject.onNext(PlayerEvent.FinishedEvent(testLibrarySource))

            verify(playQueueRepository).skipToNext()
            verify(playerCoordinatorInteractor, never()).stop(eq(PlayerType.LIBRARY))
        }

        @Test
        fun `on error event test`() {
            whenever(playQueueRepository.isCurrentCompositionAtEndOfQueue())
                .thenReturn(Single.just(false))

            playerEventSubject.onNext(PlayerEvent.ErrorEvent(mock<Exception>(), testLibrarySource))

            verify(musicProviderRepository).writeErrorAboutComposition(
                eq(CorruptionType.UNKNOWN),
                eq(playQueueItem)
            )
            verify(playQueueRepository).skipToNext()
        }

        @Test
        fun `on error event at the end of the queue`() {
            whenever(playQueueRepository.isCurrentCompositionAtEndOfQueue())
                .thenReturn(Single.just(true))

            playerEventSubject.onNext(PlayerEvent.ErrorEvent(mock<Exception>(), testLibrarySource))

            verify(playQueueRepository, never()).skipToNext()
            verify(playerCoordinatorInteractor).stop(eq(PlayerType.LIBRARY))
            verify(musicProviderRepository).writeErrorAboutComposition(
                eq(CorruptionType.UNKNOWN),
                eq(playQueueItem)
            )
        }

        @Test
        fun `on acceptable error event test`() {
            val ex: AcceptablePlayerException = mock()
            val playerException = AcceptablePlayerException(ex)

            playerEventSubject.onNext(PlayerEvent.ErrorEvent(playerException, testLibrarySource))

            verify(playerCoordinatorInteractor).error(eq(PlayerType.LIBRARY), eq(ex))
            verify(musicProviderRepository, never()).writeErrorAboutComposition(any(), any())
            verify(playQueueRepository, never()).skipToNext()
        }

        @Test
        fun `on empty queue event test`() {
            currentCompositionSubject.onNext(PlayQueueEvent(null))

            verify(playerCoordinatorInteractor).reset(eq(PlayerType.LIBRARY))
        }

    }

    @Test
    fun `skip to previous test`() {
        whenever(settingsRepository.skipConstraintMillis).thenReturn(15)
        libraryPlayerInteractor.seekTo(10L)

        libraryPlayerInteractor.skipToPrevious()
        inOrder.verify(playQueueRepository).skipToPrevious()

        libraryPlayerInteractor.seekTo(30L)
        libraryPlayerInteractor.skipToPrevious()
        inOrder.verify(playerCoordinatorInteractor).onSeekFinished(eq(0L), any())
    }
}